### ISSUE TITLE
custom claims

### ISSUE THREAD

**Issue by leantony** [jwtk/jjwt#66]
How do I add a map of custom claims to the jwt ? For example, if I wanted to add a user's role
Currently iam doing it like this

``` java
return Jwts.builder()
                .setSubject(subject)
                .setIssuer(issuer)
                .setId(UUID.randomUUID().toString())
                .setExpiration(setExpiry(expiresIn))
                .setClaims(claims)
                .signWith(this.Algo, this.getEncodedSecret())
                .compact();
```

This is just a portion of a function. Claims is variable of Map<String, Object> type. Iam using this as the custom claims, but when I do this, my token is generated without an expiry date. What am I doing wrong?

**Comment by lhazlewood** [comment]
Try setting the claims before you set any other field:

```java
return Jwts.builder()
                .setClaims(claims)
                .setSubject(subject)
                .setIssuer(issuer)
                .setId(UUID.randomUUID().toString())
                .setExpiration(setExpiry(expiresIn))
                .signWith(this.Algo, this.getEncodedSecret())
                .compact();
```

I suspect that setClaims is overwriting the other claims that have already been set.

If this works, let me know - I'll leave the ticket open to copy the map key/value pairs into the existing map (rather than replace the existing map).

**Comment by leantony** [comment]
Yes. I noticed it was overwriting the other claims. So, had to do this:
```java
return Jwts.builder()
                .claim("username", "the user's name")
                .setIssuer(issuer)
                .setExpiration(setExpiry(expiresIn))
                .setId(UUID.randomUUID().toString())
                .setSubject(subject)
                .signWith(this.Algo, this.getEncodedSecret())
                .compact();
```
I thought SetClaims() would take in my custom claims, but on reading through the source code, I saw that it wasn't doing that. But with my above method, I have to repeat the claim('key', 'value') portion, for any custom claim I have to add

**Comment by lhazlewood** [comment]
I've thought about this for a while now.

Per standard/idiomatic Java setter methods (where state is replaced with what is currently being specified) setClaims *should* replace any current claims that might be set.

That said, I think we should add a `addClaims` method that takes an input map and appends it to the existing populated claims.

**Comment by lhazlewood** [comment]
Closing as this his was resolved via #196, released in 0.8.0
